File: navalbattle-web/src/board/Cellule.js
import React from 'react';
import styled from 'styled-components';
import { getCellImage, imageDescription } from './stateDatas';

const Container = styled.td`
  width: 50px;
  height: 50px;
  border: 1px solid #5cc1ef;
  display: flex;
  cursor: ${(props) => props.isClickable && 'pointer'};
  &:hover {
    border: 1px solid ${(props) => props?.theme?.primary.main};
  }
`;
// I pass native props for not render component on change
function Cellule({
  isHit,
  isBoat,
  isDeadBoat,
  orientation,
  displayBoat,
  id,
  boardId,
  onClick,
  isClickable,
}) {
  const cell = { id, isHit, isBoat, isDeadBoat, orientation, boardId };

  if (!cell)
    return (
      <Container>
        <span>Err.</span>
      </Container>
    );

  const image = getCellImage(cell, displayBoat);

  return (
    <Container
      onClick={() => isClickable && !cell.isHit && onClick?.(cell)}
      isClickable={isClickable}
    >
      <img
        alt="water"
        title={imageDescription[image]}
        className="fullWidthImage"
        src={image}
      />
    </Container>
  );
}

export default React.memo(Cellule);

File: navalbattle-api/Models/Cellule.cs
namespace BatailleNavalGinier.Models
{
    public class Cellule
    {
        public long Id { get; set; }
        public long BoardId { get; set; }
        public int Xcoords { get; set; }
        public int Ycoords { get; set; }
        public bool IsBoat { get; set; }
        public bool IsHit { get; set; }
        public bool IsDeadBoat { get; set; }
        public string BoatIdentificator { get; set; }
        public Orientation Orientation { get; set; }

        public Board Board { get; set; }

        public Cellule(long id, long boardId, int xcoords, int ycoords, bool isBoat, bool isHit)
        {
            Id = id;
            BoardId = boardId;
            Xcoords = xcoords;
            Ycoords = ycoords;
            IsBoat = isBoat;
            IsHit = isHit;
            IsDeadBoat = false;
            Orientation = Orientation.Undefined;
        }

        public Cellule()
        {
        }

        public Cellule Copy()
        {
            return new Cellule(Id, BoardId, Xcoords, Ycoords, IsBoat, IsHit);
        }

        public void CopyProps(Cellule cell)
        {
            IsBoat = cell.IsBoat;
            BoatIdentificator = cell.BoatIdentificator;
            Orientation = cell.Orientation;
            IsDeadBoat = cell.IsDeadBoat;
            IsHit = cell.IsHit;
        }

        public void AddBoat(Orientation orientation, string boatIdentificator = "")
        {
            IsBoat = true;
            Orientation = orientation;
            BoatIdentificator = boatIdentificator;
        }
    }
}
